feat: add updateOnly option on upsert

asn6878 opened the pull request:
Closes: #11588

### Description of change
This PR introduces **fine-grained control** over update fields in `upsert` operations by adding a new `updateOnly` option to `UpsertOptions`.

Currently, `upsert` will overwrite **all provided fields on conflict**. 
This can break scenarios like authentication systems where sensitive fields (e.g., tokens) should remain unchanged during bulk imports. With the new option, developers can explicitly control which fields should be updated when a conflict occurs.

Example:
```typescript
await userRepository.upsert(userData, {
  conflictPaths: ['id'],
  updateOnly: ['email', 'username'], // Only update these fields on conflict
});


// or you can do this
await userRepository.upsert(userData, {
  conflictPaths: ['id'],
  updateOnly: {email: true, username: true},
});
```
Also following validation checks and tests have been added:
- **Unknown Column Check**: Check if any column name provided in updateOnly does not exist in the entity metadata.
- **Conflict-Update Overlap Check**: Prevents the same column from being used in both conflictPaths and updateOnly.


If you notice any potential risks with this change, or areas where additional handling might be required, I would greatly appreciate your feedback.
### Pull-Request Checklist

<!--
  Please make sure to review and check all of the following.

  If an item is not applicable, you can add "N/A" to the end.
-->

-   [x] Code is up-to-date with the `master` branch
-   [x] This pull request links relevant issues as `Fixes #11588 `
-   [x] There are new or updated unit tests validating the change
-   [ ] Documentation has been updated to reflect this change


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Upsert now supports update-only mode to control which fields are updated on conflict (array or keyed-object forms), with validation for unknown fields.
* **Bug Fixes**
  * Improved Postgres upsert behavior to reliably trigger ON CONFLICT by ensuring conflict columns are included in the INSERT column list.
* **Tests**
  * Added functional tests covering update-only behavior, validation of unknown fields, and driver-conditional execution.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

asn6878 committed:
feat: add updateOnly option on upsert

asn6878 committed:
fix: add a conflict colums to first parameter

asn6878 committed:
test: change input entity data

asn6878 committed:
fix: properly map conflict columns in upsert for postgres

asn6878 committed:
fix: apply into second parameter only when using Postgres Family

asn6878 committed:
Merge branch 'master' of https://github.com/typeorm/typeorm into add-fine-control-on-upsert

sgarner commented on test/github-issues/11588/issue-11588.ts:
```diff
@@ -0,0 +1,126 @@
+} from "../../utils/test-utils"
+import { User } from "./entity/User"
+
+describe("github issues > #11588 Fine Grained Control over Update fields on Upsert", () => {
```
We are moving away from creating tests for specific issues. Instead, please integrate tests for this feature into the relevant test suite(s) for EntityManager.

sgarner reviewed (dismissed):
@asn6878 Thanks for your contribution!

Could you please review the comments from CodeRabbit and make any necessary adjustments?

asn6878 committed:
fix: improve updateOnly validation and filtering in upsert operations

- Skip updating columns in updateOnly when their values are not provided in entities
- Add fast-fail validation for unknown columns specified in updateOnly option

asn6878 committed:
test: migrate legacy issue-dependent test into integration test

gioboa committed:
Merge branch 'master' into add-fine-control-on-upsert

asn6878 committed:
Merge branch 'master' into add-fine-control-on-upsert

asn6878 committed:
Merge branch 'master' into add-fine-control-on-upsert

asn6878 commented:
@sgarner @naorpeled  @jovanadjuric @gioboa
Hey! Would you mind taking a look at this feature when you get a chance? Thanks a lot!

asn6878 committed:
Merge branch 'master' of https://github.com/asn6878/typeorm into add-fine-control-on-upsert

asn6878 committed:
feat: validate overlap between conflictPaths and updateOnly

gioboa committed:
Merge branch 'master' into add-fine-control-on-upsert

asn6878 committed:
Merge branch 'master' into add-fine-control-on-upsert

gioboa reviewed (commented):
Can I ask you to solve the conflicts please?

asn6878 committed:
chore: merge master branch

asn6878 committed:
style: remove uninteded ts-docs

asn6878 committed:
fix: remove unintended changes

gioboa committed:
Merge branch 'master' into add-fine-control-on-upsert

gioboa committed:
Merge branch 'master' into add-fine-control-on-upsert

gioboa committed:
docs: imporve docs

gioboa reviewed (commented):
I improved the docs as well [with this commit](https://github.com/typeorm/typeorm/pull/11630/commits/26fdb0fd8ba0807af5d469c0f728e73586dd45d2). @asn6878 do you like it?

asn6878 committed:
Merge branch 'master' into add-fine-control-on-upsert

gioboa reviewed (changes requested):
is it still valid? can you solve the conflicts please?

asn6878 committed:
Merge branch 'master' of https://github.com/typeorm/typeorm into add-fine-control-on-upsert

asn6878 committed:
fix: exclude conflict columns from upsert UPDATE SET clause

asn6878 committed:
Merge branch 'add-fine-control-on-upsert' of https://github.com/asn6878/typeorm into add-fine-control-on-upsert

asn6878 committed:
test: replace connections with dataSources in upsert basic-methods tests

asn6878 committed:
fix: fallback to conflictColumns in orUpdate when overwriteColumns is empty

asn6878 committed:
Merge branch 'master' into add-fine-control-on-upsert

asn6878 committed:
Merge branch 'master' into add-fine-control-on-upsert

Commit:
Merge remote-tracking branch 'origin/master' into work-11630

# Conflicts:
#	src/entity-manager/EntityManager.ts

asn6878 committed:
Merge branch 'master' into add-fine-control-on-upsert

asn6878 committed:
Merge branch 'master' into add-fine-control-on-upsert
